010-server-side-rendering.md

```diff
@@ -0,0 +1,205 @@
+# Server-Side Rendering
+
+We should use [Next.js](https://nextjs.org) and [Heroku](https://www.heroku.com) to power server-side rendering for Phoenix.
+
+**Discussion:** https://github.com/DoSomething/rfcs/pull/10
+
+## Problem
+
+Our front-end application, Phoenix, is not as scalable, simple, or performant as we want:
+
+- Because Phoenix runs on traditional server processes, our **ability to scale is limited** (for authenticated users). If we’re unable to gracefully scale with traffic spikes, we might lose engagement at a key time.
+- The application is split between a traditional backend server, written in [Laravel](https://laravel.com), and a client-side application, written in [React](https://reactjs.org). This **increases context switching for developers**, slowing down development.
+- Because the application is almost completely rendered on the client, we see **reduced performance on slower devices** (like old phones). This may reduce engagement, since nobody likes waiting.
+- Finally, we've been limited in what we can share on social networks (whose scrapers do not execute client-side code) since any social metadata must be fetched and rendered by a separate server-side process.
```

Comment:
Playing devil's advocate, would it be fair to argue that all of these issues are addressable without resorting to server-side rendering? (e.g. completely moving our backend 3rd party API calls to front-end/GraphQL and coming up with some general strategy for SEO and scraping.) 

Is there a general argument specifically for server-side rendering notwithstanding us solving these issues on their own? (Or, is this counter-argument invalid - and these issues _are_ still big enough to warrant the switch e.g. I didn't address scaling.)

Reply:
Yeah, the first two could absolutely be solved by moving more logic to the client-side (and having Laravel simply [render that empty `app` view](https://github.com/DoSomething/phoenix-next/blob/c04752094f8234c31208c615b4f7c61dd8d8ecc6/routes/web.php#L60)). The more we do that, though, the more the second two problems suffer (rendering performance & un-scrapable social metadata).

I think what I find most appealing about tackling server-rendering is that it lets us make meaningful improvements for all of these pain points in one fell swoop.

Reply:
I think the final problem perhaps makes the cleanest case for server-side rendering.

Since Facebook & Twitter don't run client-side code, we have to provide those metatags in the server's HTML response. This means we either:
1. can't include dynamic metadata on client-rendered pages (like company/hub pages)
2. have to make all of our requests on the server, which introduces that awkward conceptual split & prevents us from using client-side routing for better performance.
3. have to double-up those requests (with a server-side Contentful API request & a client-side GraphQL request), which adds a meaningful performance penalty.

The way we'd so far tried to improve developer experience & scalability (moving to client-side rendering) has a direct negative effect on performance & "scrapability".

Reply:
Ah yeah, I'd agree that the final problem indeed made the strongest impression on me. Particularly since we've (as you well know lol) [had to deal with this](https://github.com/DoSomething/phoenix-next/pull/1979) pain point directly.

And I quite like the 'one fell swoop' point! Altogether, color me convinced. and thanks for clarifying!